File: organism-sim/src/organism_sim/world.py
from __future__ import annotations

from dataclasses import dataclass, field
from functools import lru_cache
from random import Random

import numpy as np

from .chemistry import ChemistryCatalog, Inventory, MoleculeBatch, add_batch
from .config import SimulationConfig
from .entities import Organism, Position

_MASK64 = (1 << 64) - 1


def _mix64(value: int) -> int:
    """splitmix64 finalizer: stable 64-bit scrambling for chunk seeds."""

    value = (value + 0x9E3779B97F4A7C15) & _MASK64
    value = ((value ^ (value >> 30)) * 0xBF58476D1CE4E5B9) & _MASK64
    value = ((value ^ (value >> 27)) * 0x94D049BB133111EB) & _MASK64
    return value ^ (value >> 31)


@dataclass(slots=True)
class Chunk:
    """One square region of the infinite world, generated on first touch."""

    cx: int
    cy: int
    heat: np.ndarray
    deposit_positions: set[Position] = field(default_factory=set)
    has_heat: bool = False


@dataclass(slots=True)
class World:
    """Infinite procedural world.

    The world has no edges and no wrapping. Terrain (deposits and thermal
    springs) is generated deterministically per chunk the first time the chunk
    is touched, so the world extends without bound as organisms explore it.
    Newly generated geological matter and energy are accumulated in
    ``generated_elements`` / ``generated_energy`` so conservation audits can
    separate dynamics from world generation.
    """

    config: SimulationConfig
    catalog: ChemistryCatalog
    chunk_size: int
    terrain_seed: int
    chunks: dict[tuple[int, int], Chunk] = field(default_factory=dict)
    deposits: dict[Position, Inventory] = field(default_factory=dict)
    occupancy: dict[Position, set[int]] = field(default_factory=dict)
    # Optional demand histogram over molecule IDs. When set, newly generated
    # chunks draw deposit molecules in proportion to these weights instead of
    # uniformly, mirroring the phase-1 matched-pool economy finding.
    molecule_weights: tuple[float, ...] | None = None
    generated_elements: tuple[int, ...] = ()
    generated_energy: float = 0.0

    @classmethod
    def generate(cls, config: SimulationConfig, catalog: ChemistryCatalog, rng: Random) -> World:
        world = cls(
            config=config,
            catalog=catalog,
            chunk_size=config.chunk_size,
            terrain_seed=rng.getrandbits(64),
            generated_elements=(0,) * len(catalog.elements),
        )
        # Materialize the founder region up front so the spawn area starts with
        # the configured resource density; everything else appears on demand.
        for cx in range(-(-config.width // config.chunk_size)):
            for cy in range(-(-config.height // config.chunk_size)):
                world.ensure_chunk(cx, cy)
        return world

    # ----------------------------------------------------------- chunk access

    def ensure_chunk(self, cx: int, cy: int) -> Chunk:
        chunk = self.chunks.get((cx, cy))
        if chunk is not None:
            return chunk
        chunk = Chunk(cx, cy, np.zeros((self.chunk_size, self.chunk_size), dtype=np.float64))
        self.chunks[(cx, cy)] = chunk
        self._generate_chunk(chunk)
        return chunk

    def chunk_at(self, position: Position) -> Chunk | None:
        """Read-only chunk lookup; never generates terrain."""

        return self.chunks.get((position[0] // self.chunk_size, position[1] // self.chunk_size))

    def ensure_positions(self, positions) -> None:
        size = self.chunk_size
        keys = {(x // size, y // size) for x, y in positions}
        for cx, cy in keys:
            self.ensure_chunk(cx, cy)

    def _chunk_rng(self, chunk: Chunk) -> Random:
        seed = _mix64(self.terrain_seed ^ _mix64(chunk.cx) ^ (_mix64(chunk.cy) * 3)) & _MASK64
        return Random(seed)

    def _generate_chunk(self, chunk: Chunk) -> None:
        config = self.config
        rng = self._chunk_rng(chunk)
        size = self.chunk_size
        chunk_area = size * size
        region_area = max(1, config.width * config.height)
        density = config.initial_deposits / region_area
        expected = density * chunk_area
        count = int(expected) + (1 if rng.random() < expected % 1.0 else 0)
        elements = [0] * len(self.catalog.elements)
        energy = 0.0
        for _ in range(count):
            x = chunk.cx * size + rng.randrange(size)
            y = chunk.cy * size + rng.randrange(size)
            molecule = (
                self._weighted_molecule(rng) if self.molecule_weights is not None
                else rng.choice(self.catalog.molecules)
            )
            units = rng.randint(config.initial_batch_min, config.initial_batch_max)
            charge = rng.uniform(0.45, 1.0)
            batch = MoleculeBatch(molecule.molecule_id, units, molecule.energy_capacity * units * charge)
            self.deposit_batch((x, y), batch)
            energy += batch.energy
            for element_id, amount in enumerate(molecule.composition):
                elements[element_id] += amount * units
        # Thermal springs: same density scaling as the original whole-map seeding.
        spring_expected = max(1, config.initial_deposits // 8) * chunk_area / region_area
        springs = int(spring_expected) + (1 if rng.random() < spring_expected % 1.0 else 0)
        for _ in range(springs):
            x = chunk.cx * size + rng.randrange(size)
            y = chunk.cy * size + rng.randrange(size)
            amount = rng.uniform(0.1, 2.0)
            self.add_heat((x, y), amount)
            energy += amount
        self.generated_elements = tuple(a + b for a, b in zip(self.generated_elements, elements))
        self.generated_energy += energy

    # ------------------------------------------------------------ coordinates

    def _weighted_molecule(self, rng: Random):
        weights = self.molecule_weights or ()
        draw = rng.random() * sum(weights)
        cumulative = 0.0
        for molecule, weight in zip(self.catalog.molecules, weights):
            cumulative += weight
            if draw < cumulative:
                return molecule
        return self.catalog.molecules[-1]

    def index(self, position: Position) -> Position:
        """Canonical deposit/occupancy key: absolute (unwrapped) coordinates."""

        return (position[0], position[1])

    def position(self, key: Position) -> Position:
        return (key[0], key[1])

    def delta(self, source: Position, target: Position) -> Position:
        return target[0] - source[0], target[1] - source[1]

    def distance(self, source: Position, target: Position) -> int:
        dx, dy = self.delta(source, target)
        return max(abs(dx), abs(dy))

    @staticmethod
    @lru_cache(maxsize=64)
    def footprint_offsets(area: int) -> tuple[Position, ...]:
        radius = 0
        candidates: list[Position] = []
        while len(candidates) < area:
            candidates = [
                (dx, dy)
                for dy in range(-radius, radius + 1)
                for dx in range(-radius, radius + 1)
            ]
            candidates.sort(key=lambda item: (item[0] * item[0] + item[1] * item[1], abs(item[1]), abs(item[0]), item))
            radius += 1
        return tuple(candidates[:area])

    @classmethod
    @lru_cache(maxsize=512)
    def nearby_offsets(cls, area: int, radius: int) -> tuple[Position, ...]:
        offsets = {
            (fx + dx, fy + dy)
            for fx, fy in cls.footprint_offsets(area)
            for dy in range(-radius, radius + 1)
            for dx in range(-radius, radius + 1)
        }
        return tuple(sorted(offsets))

    def footprint(self, organism: Organism) -> tuple[Position, ...]:
        origin_x, origin_y = organism.position
        area = organism.area(self.catalog)
        key = (origin_x, origin_y, area)
        if organism._footprint_key != key:
            organism._footprint_key = key
            organism._footprint_cells = tuple(
                (origin_x + dx, origin_y + dy) for dx, dy in self.footprint_offsets(area)
            )
        self.ensure_positions(organism._footprint_cells)
        return organism._footprint_cells

    def rebuild_occupancy(self, organisms: dict[int, Organism]) -> None:
        self.occupancy.clear()
        for organism in organisms.values():
            if organism.alive:
                self.add_to_occupancy(organism)

    def add_to_occupancy(self, organism: Organism) -> None:
        """Add one organism's current footprint to the spatial index."""

        for position in self.footprint(organism):
            self.occupancy.setdefault(position, set()).add(organism.organism_id)

    def remove_from_occupancy(self, organism: Organism) -> None:
        """Remove one organism's current footprint from the spatial index."""

        self.remove_id_from_occupancy(organism.organism_id, self.footprint(organism))

    def remove_id_from_occupancy(self, organism_id: int, positions: tuple[Position, ...]) -> None:
        """Remove an ID from known positions, including a dead organism's old footprint."""

        for position in positions:
            occupants = self.occupancy.get(position)
            if occupants is None:
                continue
            occupants.discard(organism_id)
            if not occupants:
                del self.occupancy[position]

    def organism_ids_at(self, position: Position) -> set[int]:
        return self.occupancy.get((position[0], position[1]), set())

    def can_place(self, organism: Organism, position: Position, ignore: set[int] | None = None) -> bool:
        origin_x, origin_y = position
        area = organism.area(self.catalog)
        offsets = self.footprint_offsets(area)
        size = self.chunk_size
        chunk_keys = {
            ((origin_x + dx) // size, (origin_y + dy) // size) for dx, dy in offsets
        }
        for cx, cy in chunk_keys:
            self.ensure_chunk(cx, cy)
        occupancy = self.occupancy
        if ignore is None:
            for dx, dy in offsets:
                if occupancy.get((origin_x + dx, origin_y + dy)) is not None:
                    return False
            return True
        for dx, dy in offsets:
            occupants = occupancy.get((origin_x + dx, origin_y + dy))
            if occupants is None:
                continue
            for organism_id in occupants:
                if organism_id not in ignore:
                    return False
        return True

    def nearby_cells(self, organism: Organism, radius: int) -> tuple[Position, ...]:
        origin_x, origin_y = organism.position
        area = organism.area(self.catalog)
        key = (origin_x, origin_y, area)
        cached = organism._nearby_cells_cache.get(radius)
        if cached is None or cached[0] != key:
            cells = tuple(
                (origin_x + dx, origin_y + dy) for dx, dy in self.nearby_offsets(area, radius)
            )
            cached = (key, cells)
            organism._nearby_cells_cache[radius] = cached
        self.ensure_positions(cached[1])
        return cached[1]

    def nearby_organism_ids(self, organism: Organism, radius: int) -> set[int]:
        result: set[int] = set()
        for cell in self.nearby_cells(organism, radius):
            occupants = self.occupancy.get(cell)
            if occupants is not None:
                result.update(occupants)
        result.discard(organism.organism_id)
        return result

    # -------------------------------------------------------------- deposits

    def deposit_batch(self, position: Position, batch: MoleculeBatch) -> None:
        if batch.count <= 0:
            return
        position = (position[0], position[1])
        inventory = self.deposits.setdefault(position, {})
        add_batch(inventory, batch)
        chunk = self.ensure_chunk(position[0] // self.chunk_size, position[1] // self.chunk_size)
        chunk.deposit_positions.add(position)

    def remove_deposit(self, position: Position) -> None:
        position = (position[0], position[1])
        self.deposits.pop(position, None)
        chunk = self.chunks.get((position[0] // self.chunk_size, position[1] // self.chunk_size))
        if chunk is not None:
            chunk.deposit_positions.discard(position)

    def inventory_at(self, position: Position) -> Inventory:
        position = (position[0], position[1])
        inventory = self.deposits.setdefault(position, {})
        if inventory:
            chunk = self.ensure_chunk(position[0] // self.chunk_size, position[1] // self.chunk_size)
            chunk.deposit_positions.add(position)
        return inventory

    def nonempty_inventory_at(self, position: Position) -> Inventory | None:
        return self.deposits.get((position[0], position[1]))

    # ----------------------------------------------------------------- heat

    def clear_heat(self) -> None:
        for chunk in self.chunks.values():
            chunk.heat.fill(0.0)
            chunk.has_heat = False

    def add_heat(self, position: Position, amount: float) -> None:
        if amount <= 0.0:
            return
        size = self.chunk_size
        chunk = self.ensure_chunk(position[0] // size, position[1] // size)
        chunk.heat[position[1] % size, position[0] % size] += amount
        chunk.has_heat = True

    def heat_at(self, position: Position) -> float:
        size = self.chunk_size
        chunk = self.chunks.get((position[0] // size, position[1] // size))
        if chunk is None:
            return 0.0
        return float(chunk.heat[position[1] % size, position[0] % size])

    def remove_heat(self, positions: tuple[Position, ...], amount: float) -> float:
        if amount <= 0.0:
            return 0.0
        size = self.chunk_size
        if len(positions) == 1:
            x, y = positions[0]
            chunk = self.ensure_chunk(x // size, y // size)
            lx, ly = x % size, y % size
            available = float(chunk.heat[ly, lx])
            removed = min(amount, available)
            if removed <= 0.0:
                return 0.0
            remaining = removed
            share = removed * float(chunk.heat[ly, lx]) / available
            debit = min(float(chunk.heat[ly, lx]), share, remaining)
            chunk.heat[ly, lx] -= debit
            remaining -= debit
            if remaining > 1e-12:
                debit = min(float(chunk.heat[ly, lx]), remaining)
                chunk.heat[ly, lx] -= debit
                remaining -= debit
            return removed - max(0.0, remaining)
        self.ensure_positions(positions)
        chunks = [
            (self.chunks[(x // size, y // size)], x % size, y % size) for x, y in positions
        ]
        available = sum(float(chunk.heat[ly, lx]) for chunk, lx, ly in chunks)
        removed = min(amount, available)
        if removed <= 0.0:
            return 0.0
        remaining = removed
        for chunk, lx, ly in chunks:
            share = removed * float(chunk.heat[ly, lx]) / available
            debit = min(float(chunk.heat[ly, lx]), share, remaining)
            chunk.heat[ly, lx] -= debit
            remaining -= debit
        if remaining > 1e-12:
            chunk, lx, ly = chunks[0]
            debit = min(float(chunk.heat[ly, lx]), remaining)
            chunk.heat[ly, lx] -= debit
            remaining -= debit
        return removed - max(0.0, remaining)

    def diffuse_heat(self) -> None:
        """Diffuse heat chunk by chunk with no-flux edges toward ungenerated space.

        Every cell keeps the flux that would leave toward a missing neighbour
        chunk, so energy is conserved exactly no matter which chunks exist.
        """

        rate = self.config.heat_diffusion
        if rate <= 0.0:
            return
        active: dict[tuple[int, int], Chunk] = {}
        for key, chunk in self.chunks.items():
            if not chunk.has_heat:
                continue
            active[key] = chunk
            cx, cy = key
            for neighbor_key in ((cx + 1, cy), (cx - 1, cy), (cx, cy + 1), (cx, cy - 1)):
                neighbor = self.chunks.get(neighbor_key)
                if neighbor is not None:
                    active.setdefault(neighbor_key, neighbor)
        if not active:
            return
        sources = {key: chunk.heat.copy() for key, chunk in active.items()}
        for key, chunk in active.items():
            s = sources[key]
            new = (1.0 - 4.0 * rate) * s
            new[1:, :] += rate * s[:-1, :]
            new[:-1, :] += rate * s[1:, :]
            new[:, 1:] += rate * s[:, :-1]
            new[:, :-1] += rate * s[:, 1:]
            cx, cy = key
            up = sources.get((cx, cy - 1))
            new[0, :] += rate * (up[-1, :] if up is not None else s[0, :])
            down = sources.get((cx, cy + 1))
            new[-1, :] += rate * (down[0, :] if down is not None else s[-1, :])
            left = sources.get((cx - 1, cy))
            new[:, 0] += rate * (left[:, -1] if left is not None else s[:, 0])
            right = sources.get((cx + 1, cy))
            new[:, -1] += rate * (right[:, 0] if right is not None else s[:, -1])
            np.clip(new, 0.0, None, out=new)
            chunk.heat = new
            chunk.has_heat = bool(new.any())

    def total_heat(self) -> float:
        return float(sum(chunk.heat.sum(dtype=np.float64) for chunk in self.chunks.values() if chunk.has_heat))

File: organism-sim/rust/src/world.rs
//! Chunked infinite world: per-chunk heat fields, deposit inventories,
//! occupancy, and on-demand procedural chunk generation. Port of
//! `organism_sim.world` (chunk RNG is the kernel's own xoshiro256++).

use crate::chemistry::{add_batch, Batch, Inventory};
use crate::rng::{mix64, Rng};
use rustc_hash::FxHashMap;

pub type Position = (i64, i64);

#[derive(Clone, Debug)]
pub struct Chunk {
    pub heat: Vec<f64>,
    pub occupancy: Vec<Vec<u32>>,
    pub has_heat: bool,
}

/// Molecule facts the world generator needs (avoids holding a &Catalog).
#[derive(Clone, Debug)]
pub struct GenMolecule {
    pub molecule_id: u16,
    pub energy_capacity: f64,
    pub composition: Vec<u8>,
}

#[derive(Clone)]
pub struct WorldGenParams {
    pub region_width: i64,
    pub region_height: i64,
    pub chunk_size: usize,
    pub initial_deposits: i64,
    pub batch_min: i64,
    pub batch_max: i64,
    pub element_count: usize,
    /// Optional demand histogram over molecule ids (deposit_match_ecology).
    pub molecule_weights: Option<Vec<f64>>,
}

pub struct World {
    pub params: WorldGenParams,
    pub terrain_seed: u64,
    pub chunks: FxHashMap<(i64, i64), Chunk>,
    pub deposits: FxHashMap<Position, Inventory>,
    /// Density values are divided by `biodeposit_scale`, making uniform decay
    /// O(1) per tick instead of a scan over every accumulated site.
    pub biodeposits: FxHashMap<Position, f64>,
    pub biodeposit_scale: f64,
    pub occupied_cells: usize,
    pub generated_elements: Vec<i64>,
    pub generated_energy: f64,
    molecules: Vec<GenMolecule>,
}

impl World {
    pub fn generate(
        params: WorldGenParams,
        molecules: Vec<GenMolecule>,
        terrain_seed: u64,
    ) -> World {
        let element_count = params.element_count;
        let mut world = World {
            params,
            terrain_seed,
            chunks: FxHashMap::default(),
            deposits: FxHashMap::default(),
            biodeposits: FxHashMap::default(),
            biodeposit_scale: 1.0,
            occupied_cells: 0,
            generated_elements: vec![0; element_count],
            generated_energy: 0.0,
            molecules,
        };
        let cs = world.params.chunk_size as i64;
        let cx_max = -(-world.params.region_width / cs);
        let cy_max = -(-world.params.region_height / cs);
        for cx in 0..cx_max {
            for cy in 0..cy_max {
                world.ensure_chunk(cx, cy);
            }
        }
        world
    }

    pub fn chunk_rng(&self, cx: i64, cy: i64) -> Rng {
        let seed = mix64(self.terrain_seed ^ mix64(cx as u64) ^ mix64(cy as u64).wrapping_mul(3));
        Rng::new(seed)
    }

    pub fn ensure_chunk(&mut self, cx: i64, cy: i64) {
        if self.chunks.contains_key(&(cx, cy)) {
            return;
        }
        let size = self.params.chunk_size;
        self.chunks.insert(
            (cx, cy),
            Chunk {
                heat: vec![0.0; size * size],
                occupancy: vec![Vec::new(); size * size],
                has_heat: false,
            },
        );
        self.generate_chunk(cx, cy);
    }

    fn generate_chunk(&mut self, cx: i64, cy: i64) {
        let mut rng = self.chunk_rng(cx, cy);
        let size = self.params.chunk_size as i64;
        let chunk_area = size * size;
        let region_area = (self.params.region_width * self.params.region_height).max(1);
        let density = self.params.initial_deposits as f64 / region_area as f64;
        let expected = density * chunk_area as f64;
        let count = expected as i64 + if rng.chance(expected % 1.0) { 1 } else { 0 };
        let mut elements = vec![0i64; self.params.element_count];
        let mut energy = 0.0f64;
        for _ in 0..count {
            let x = cx * size + rng.below(size as u64) as i64;
            let y = cy * size + rng.below(size as u64) as i64;
            let molecule = self.pick_molecule(&mut rng);
            let units = rng.randint(self.params.batch_min, self.params.batch_max);
            let charge = rng.uniform(0.45, 1.0);
            let batch = Batch {
                molecule_id: molecule.molecule_id,
                count: units,
                energy: molecule.energy_capacity * units as f64 * charge,
            };
            energy += batch.energy;
            for (e, &amount) in molecule.composition.iter().enumerate() {
                elements[e] += amount as i64 * units;
            }
            self.deposit_batch_raw((x, y), batch);
        }
        let spring_expected = (self.params.initial_deposits / 8).max(1) as f64 * chunk_area as f64
            / region_area as f64;
        let springs = spring_expected as i64
            + if rng.chance(spring_expected % 1.0) {
                1
            } else {
                0
            };
        for _ in 0..springs {
            let x = cx * size + rng.below(size as u64) as i64;
            let y = cy * size + rng.below(size as u64) as i64;
            let amount = rng.uniform(0.1, 2.0);
            self.add_heat((x, y), amount);
            energy += amount;
        }
        for (a, b) in self.generated_elements.iter_mut().zip(elements.iter()) {
            *a += *b;
        }
        self.generated_energy += energy;
    }

    fn pick_molecule(&self, rng: &mut Rng) -> &GenMolecule {
        match &self.params.molecule_weights {
            None => &self.molecules[rng.below(self.molecules.len() as u64) as usize],
            Some(weights) => {
                let total: f64 = weights.iter().sum();
                let mut draw = rng.f64() * total;
                let mut cumulative = 0.0;
                for (molecule, weight) in self.molecules.iter().zip(weights.iter()) {
                    cumulative += *weight;
                    if draw < cumulative {
                        return molecule;
                    }
                }
                // Guard against zero total weight.
                if total <= 0.0 {
                    draw = rng.f64() * self.molecules.len() as f64;
                    return &self.molecules[draw.floor() as usize % self.molecules.len()];
                }
                self.molecules.last().unwrap()
            }
        }
    }

    pub fn set_molecule_weights(&mut self, weights: Vec<f64>) {
        self.params.molecule_weights = Some(weights);
    }

    // ------------------------------------------------------------ coordinates

    #[inline]
    pub fn distance(source: Position, target: Position) -> i64 {
        let dx = target.0 - source.0;
        let dy = target.1 - source.1;
        dx.abs().max(dy.abs())
    }

    #[inline]
    pub fn chunk_of(&self, position: Position) -> (i64, i64) {
        let size = self.params.chunk_size as i64;
        (position.0.div_euclid(size), position.1.div_euclid(size))
    }

    #[inline]
    fn local_index(&self, position: Position) -> usize {
        let size = self.params.chunk_size as i64;
        let lx = position.0.rem_euclid(size);
        let ly = position.1.rem_euclid(size);
        (ly * size + lx) as usize
    }

    pub fn ensure_positions(&mut self, positions: &[Position]) {
        for &(x, y) in positions {
            let key = self.chunk_of((x, y));
            self.ensure_chunk(key.0, key.1);
        }
    }

    // -------------------------------------------------------------- occupancy

    pub fn add_to_occupancy(&mut self, organism_id: u32, positions: &[Position]) {
        self.ensure_positions(positions);
        for &position in positions {
            let key = self.chunk_of(position);
            let index = self.local_index(position);
            let occupants = &mut self.chunks.get_mut(&key).unwrap().occupancy[index];
            if !occupants.contains(&organism_id) {
                if occupants.is_empty() {
                    self.occupied_cells += 1;
                }
                occupants.push(organism_id);
            }
        }
    }

    pub fn remove_id_from_occupancy(&mut self, organism_id: u32, positions: &[Position]) {
        for &position in positions {
            let key = self.chunk_of(position);
            let index = self.local_index(position);
            if let Some(chunk) = self.chunks.get_mut(&key) {
                let occupants = &mut chunk.occupancy[index];
                let was_occupied = !occupants.is_empty();
                occupants.retain(|&id| id != organism_id);
                if was_occupied && occupants.is_empty() {
                    self.occupied_cells -= 1;
                }
            }
        }
    }

    pub fn occupants_at(&self, position: Position) -> Option<&[u32]> {
        let key = self.chunk_of(position);
        let chunk = self.chunks.get(&key)?;
        let occupants = &chunk.occupancy[self.local_index(position)];
        if occupants.is_empty() {
            None
        } else {
            Some(occupants)
        }
    }

    /// True if no organism other than `ignore` occupies any of `positions`.
    /// (Mirrors World.can_place; footprint positions are precomputed.)
    pub fn area_free(&mut self, positions: &[Position], ignore: u32) -> bool {
        self.ensure_positions(positions);
        for &position in positions {
            if self
                .occupants_at(position)
                .is_some_and(|occupants| occupants.iter().any(|&id| id != ignore))
            {
                return false;
            }
        }
        true
    }

    // --------------------------------------------------------------- deposits

    pub fn deposit_batch_raw(&mut self, position: Position, batch: Batch) {
        if batch.count <= 0 {
            return;
        }
        let key = self.chunk_of(position);
        self.ensure_chunk(key.0, key.1);
        let inventory = self.deposits.entry(position).or_default();
        add_batch(inventory, batch);
    }

    pub fn remove_deposit(&mut self, position: Position) {
        self.deposits.remove(&position);
    }

    #[inline]
    pub fn biodeposit_density_at(&self, position: Position) -> f64 {
        self.biodeposits.get(&position).copied().unwrap_or(0.0) * self.biodeposit_scale
    }

    pub fn add_biodeposit(&mut self, position: Position, structural_mass: f64) {
        if structural_mass > 0.0 {
            *self.biodeposits.entry(position).or_default() +=
                structural_mass / self.biodeposit_scale;
        }
    }

    pub fn erode_biodeposit(&mut self, position: Position, structural_mass: f64) {
        let Some(density) = self.biodeposits.get_mut(&position) else {
            return;
        };
        *density = (*density - structural_mass.max(0.0) / self.biodeposit_scale).max(0.0);
        if *density * self.biodeposit_scale <= 1.0e-9 {
            self.biodeposits.remove(&position);
        }
    }

    pub fn decay_biodeposits(&mut self, rate: f64) {
        if rate <= 0.0 {
            return;
        }
        self.biodeposit_scale *= (1.0 - rate).clamp(0.0, 1.0);
        if self.biodeposit_scale < 1.0e-100 {
            let scale = self.biodeposit_scale;
            self.biodeposits.retain(|_, density| {
                *density *= scale;
                *density > 1.0e-9
            });
            self.biodeposit_scale = 1.0;
        }
    }

    pub fn total_biodeposit_density(&self) -> f64 {
        self.biodeposits.values().sum::<f64>() * self.biodeposit_scale
    }

    // ------------------------------------------------------------------- heat

    pub fn add_heat(&mut self, position: Position, amount: f64) {
        if amount <= 0.0 {
            return;
        }
        let key = self.chunk_of(position);
        self.ensure_chunk(key.0, key.1);
        let idx = self.local_index(position);
        let chunk = self.chunks.get_mut(&key).unwrap();
        chunk.heat[idx] += amount;
        chunk.has_heat = true;
    }

    pub fn heat_at_peek(&self, position: Position) -> f64 {
        let key = self.chunk_of(position);
        match self.chunks.get(&key) {
            Some(chunk) => chunk.heat[self.local_index(position)],
            None => 0.0,
        }
    }

    pub fn remove_heat_at(&mut self, position: Position, amount: f64) -> f64 {
        let key = self.chunk_of(position);
        self.ensure_chunk(key.0, key.1);
        let index = self.local_index(position);
        let available = self.chunks[&key].heat[index];
        let removed = amount.max(0.0).min(available);
        if removed <= 0.0 {
            return 0.0;
        }
        // Preserve the single-cell arithmetic of remove_heat.
        let share = removed * available / available;
        let debit = available.min(share).min(removed);
        self.chunks.get_mut(&key).unwrap().heat[index] -= debit;
        debit
    }

    pub fn remove_heat(&mut self, positions: &[Position], amount: f64) -> f64 {
        self.ensure_positions(positions);
        let mut cells: Vec<((i64, i64), usize, f64)> = Vec::with_capacity(positions.len());
        let mut available = 0.0f64;
        for &position in positions {
            let key = self.chunk_of(position);
            let idx = self.local_index(position);
            let cell_heat = self.chunks.get(&key).unwrap().heat[idx];
            available += cell_heat;
            cells.push((key, idx, cell_heat));
        }
        let removed = amount.max(0.0).min(available);
        if removed <= 0.0 {
            return 0.0;
        }
        let mut remaining = removed;
        for (key, idx, cell_heat) in &cells {
            let share = removed * *cell_heat / available;
            let debit = cell_heat.min(share).min(remaining);
            let chunk = self.chunks.get_mut(key).unwrap();
            chunk.heat[*idx] -= debit;
            remaining -= debit;
        }
        if remaining > 1e-12 {
            let (key, idx, _) = cells[0];
            let chunk = self.chunks.get_mut(&key).unwrap();
            let debit = chunk.heat[idx].min(remaining);
            chunk.heat[idx] -= debit;
            remaining -= debit;
        }
        removed - remaining.max(0.0)
    }

    /// Chunk-by-chunk heat diffusion with no-flux edges toward ungenerated
    /// space. Exactly mirrors World.diffuse_heat.
    pub fn diffuse_heat(
        &mut self,
        rate: f64,
        sources: &mut Vec<((i64, i64), Vec<f64>)>,
        source_index: &mut FxHashMap<(i64, i64), usize>,
        scratch: &mut Vec<f64>,
    ) {
        if rate <= 0.0 {
            return;
        }
        // Active set: chunks with heat plus their existing 4-neighbours.
        let mut active: rustc_hash::FxHashSet<(i64, i64)> = rustc_hash::FxHashSet::default();
        let keys: Vec<(i64, i64)> = self
            .chunks
            .iter()
            .filter(|(_, c)| c.has_heat)
            .map(|(k, _)| *k)
            .collect();
        for &(cx, cy) in &keys {
            active.insert((cx, cy));
            for nk in [(cx + 1, cy), (cx - 1, cy), (cx, cy + 1), (cx, cy - 1)] {
                if self.chunks.contains_key(&nk) {
                    active.insert(nk);
                }
            }
        }
        if active.is_empty() {
            return;
        }
        sources.clear();
        source_index.clear();
        let size = self.params.chunk_size;
        for &key in &active {
            source_index.insert(key, sources.len());
            let heat = self.chunks[&key].heat.clone();
            sources.push((key, heat));
        }
        scratch.clear();
        scratch.resize(size * size, 0.0);
        for source_position in 0..sources.len() {
            let (key, src) = &sources[source_position];
            let (cx, cy) = *key;
            let up = source_index
                .get(&(cx, cy - 1))
                .map(|&index| &sources[index].1);
            let down = source_index
                .get(&(cx, cy + 1))
                .map(|&index| &sources[index].1);
            let left = source_index
                .get(&(cx - 1, cy))
                .map(|&index| &sources[index].1);
            let right = source_index
                .get(&(cx + 1, cy))
                .map(|&index| &sources[index].1);
            for ly in 0..size {
                for lx in 0..size {
                    let s = src[ly * size + lx];
                    let mut value = (1.0 - 4.0 * rate) * s;
                    // in-chunk neighbours with no-flux fallback at map edge
                    let up_v = if ly > 0 {
                        src[(ly - 1) * size + lx]
                    } else {
                        up.map(|u| u[(size - 1) * size + lx]).unwrap_or(s)
                    };
                    let down_v = if ly + 1 < size {
                        src[(ly + 1) * size + lx]
                    } else {
                        down.map(|d| d[lx]).unwrap_or(s)
                    };
                    let left_v = if lx > 0 {
                        src[ly * size + lx - 1]
                    } else {
                        left.map(|l| l[ly * size + size - 1]).unwrap_or(s)
                    };
                    let right_v = if lx + 1 < size {
                        src[ly * size + lx + 1]
                    } else {
                        right.map(|r| r[ly * size]).unwrap_or(s)
                    };
                    value += rate * (up_v + down_v + left_v + right_v);
                    scratch[ly * size + lx] = value.max(0.0);
                }
            }
            let chunk = self.chunks.get_mut(key).unwrap();
            chunk.heat.copy_from_slice(scratch);
            chunk.has_heat = chunk.heat.iter().any(|&v| v != 0.0);
        }
    }

    pub fn total_heat(&self) -> f64 {
        let mut total = 0.0f64;
        for chunk in self.chunks.values() {
            if chunk.has_heat {
                total += chunk.heat.iter().sum::<f64>();
            }
        }
        total
    }

    pub fn deposit_energy(&self) -> f64 {
        let mut total = 0.0;
        for inventory in self.deposits.values() {
            for batch in inventory {
                total += batch.energy;
            }
        }
        total
    }
}
